eppo_core/src/event_ingestion/delivery.rs

```diff
@@ -50,108 +46,88 @@ impl DeliveryStatus {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+pub(super) struct DeliveryConfig {
+    pub max_retries: u32,
+    pub base_retry_delay: Duration,
+    pub max_retry_delay: Duration,
+}
+
 pub(super) async fn delivery(
-    mut uplink: mpsc::Receiver<BatchedMessage<QueuedEvent>>,
-    retry_downlink: mpsc::Sender<BatchedMessage<QueuedEvent>>,
+    mut uplink: mpsc::Receiver<BatchedMessage<Event>>,
     delivery_status: mpsc::Sender<DeliveryStatus>,
-    max_retries: u32,
     event_delivery: EventDelivery,
+    config: DeliveryConfig,
 ) -> Option<()> {
+    // We use this unbounded channel to loop back messages that need retrying.
+    let (retries_tx, mut retries_rx) =
+        mpsc::unbounded_channel::<(/* attempts: */ u32, BatchedMessage<Event>)>();
+
     loop {
-        let BatchedMessage {
-            batch,
-            flush: _flush,
-        } = uplink.recv().await?;
-        if batch.is_empty() {
-            continue;
+        // Randomly select between sending a new batch or retrying an old one.
+        let (attempts, msg) = tokio::select! {
+            msg = uplink.recv() => (0, msg?),
+            msg = retries_rx.recv() => msg?,
+        };
+
+        let BatchedMessage { batch, flush } = msg;
+
+        let mut result = event_delivery.deliver(batch).await;
+
+        if attempts >= config.max_retries {
+            // Exceeded max retries -> promote retriable errors to permanent ones.
+            result.failure.append(&mut result.retry);
```

Comment:
`append()` removes all events from `result.retry` and inserts them into `result.failure`.